File: backend/src/main/java/com/coing/domain/coin/ticker/service/TickerService.java
package com.coing.domain.coin.ticker.service;

import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.messaging.simp.SimpMessageSendingOperations;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import com.coing.domain.coin.market.entity.Market;
import com.coing.domain.coin.market.service.MarketService;
import com.coing.domain.coin.ticker.dto.TickerDto;
import com.coing.domain.coin.ticker.entity.Ticker;
import com.coing.global.exception.BusinessException;
import com.coing.infra.upbit.dto.UpbitApiTradeDto;
import com.coing.util.MessageUtil;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class TickerService {

	@Value("${upbit.trade.uri}")
	private String UPBIT_TRADE_URI;

	private final MessageUtil messageUtil;
	private final MarketService marketService;
	private final SimpMessageSendingOperations messagingTemplate;
	private final RestTemplate restTemplate;
	private final Map<String, TickerDto> tickerCache = new ConcurrentHashMap<>();
	private final Map<String, Long> lastSentTime = new ConcurrentHashMap<>();
	private static final long THROTTLE_INTERVAL_MS = 200;

	public TickerDto getTicker(String market) {
		return Optional.ofNullable(tickerCache.get(market)).orElseThrow(() -> new BusinessException(
			messageUtil.resolveMessage("ticker.not.found"),
			HttpStatus.NOT_FOUND));
	}

	public List<TickerDto> getTickers(List<String> markets) {
		return tickerCache.entrySet()
			.stream()
			.filter(e -> markets.contains(e.getKey()))
			.map(Map.Entry::getValue)
			.collect(Collectors.toList());
	}

	public void updateTicker(Ticker ticker) {
		Market market = marketService.getCachedMarketByCode(ticker.getCode());
		TickerDto dto = TickerDto.from(ticker, market);
		tickerCache.put(ticker.getCode(), dto);
		publish(dto);
	}

	/**
	 * 단기 변동률(1분) 계산
	 */
	public double calculateOneMinuteRate(String code, double currentTradePrice) {
		double pastPrice = fetchPastTradePrice(code, 1, 1);
		return (currentTradePrice - pastPrice) / pastPrice;
	}

	/**
	 * n분전 시간 계산
	 */
	private LocalTime getPastTime(int minute) {
		return LocalTime.now().minusMinutes(minute);
	}

	/**
	 * 체결가 조회 API 활용하여 n분 전 가격 조회
	 */
	private Double fetchPastTradePrice(String code, int minute, int count) {
		String pastTime = getPastTime(minute).format(DateTimeFormatter.ofPattern("HH:mm:ss"));
		String url = String.format("%s?market=%s&to=%s&count=%d", UPBIT_TRADE_URI, code, pastTime, count);

		ResponseEntity<UpbitApiTradeDto[]> response = restTemplate.getForEntity(url,
			UpbitApiTradeDto[].class);

		if (!response.getStatusCode().is2xxSuccessful() || response.getBody() == null) {
			throw new BusinessException(messageUtil.resolveMessage("trade.fetch.failed"),
				HttpStatus.INTERNAL_SERVER_ERROR);
		}

		List<UpbitApiTradeDto> trades = Arrays.asList(response.getBody());

		return trades.getFirst().getTradePrice();
	}

	/**
	 * WebSocket을 통해 실시간 Ticker 데이터 publish
	 */
	public void publish(TickerDto dto) {
		String market = dto.code();
		long now = System.currentTimeMillis();
		long lastSent = lastSentTime.getOrDefault(market, 0L);

		if (now - lastSent >= THROTTLE_INTERVAL_MS) {
			messagingTemplate.convertAndSend("/sub/coin/ticker/" + market, dto);
			lastSentTime.put(market, now);
		}
	}
}

File: backend/src/main/java/com/coing/domain/coin/ticker/service/TickerService.kt
package com.coing.domain.coin.ticker.service

import com.coing.domain.coin.common.port.CoinDataHandler
import com.coing.domain.coin.common.port.EventPublisher
import com.coing.domain.coin.market.service.MarketService
import com.coing.domain.coin.ticker.dto.TickerDto
import com.coing.domain.coin.ticker.entity.Ticker
import com.coing.domain.notification.entity.OneMinuteRate
import com.coing.domain.notification.service.PushService
import com.coing.global.exception.BusinessException
import com.coing.util.MessageUtil
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentLinkedDeque

@Service
class TickerService(
    private val messageUtil: MessageUtil,
    private val marketService: MarketService,
    private val eventPublisher: EventPublisher<TickerDto>,
    private val pushService: PushService,
    private val coroutineScope: CoroutineScope = CoroutineScope(Dispatchers.IO)
) : CoinDataHandler<Ticker> {
    private val tickerListCache: MutableMap<String, Deque<TickerDto>> = ConcurrentHashMap()
    private val lastSentTime = ConcurrentHashMap<String, Long>()
    private val lastPushSentTime = ConcurrentHashMap<String, Long>()
    private val pushMutexMap = ConcurrentHashMap<String, Mutex>()

    companion object {
        private const val THROTTLE_INTERVAL_MS = 200L
        private const val SLIDING_WINDOW_MS = 60_000L
        private const val PUSH_THROTTLE_INTERVAL_MS = 60_000L
    }

    override fun update(data: Ticker) {
        val code = data.code
        val dto = TickerDto.from(data, marketService.getCachedMarketByCode(code))
        updateTickerListCache(code, dto)
        publish(dto)
        pushMessage(dto)
    }

    override fun fallbackUpdate(lastUpdate: String) {
        tickerListCache.forEach { (code, deque) ->
            val latest = deque.peekLast()
            if (latest != null) {
                latest.isFallback = true
                latest.lastUpdate = lastUpdate
                publish(latest)
            }
        }

    }

    fun getTicker(market: String): TickerDto {
        val deque = tickerListCache[market]
        if (deque.isNullOrEmpty()) {
            throw BusinessException(messageUtil.resolveMessage("ticker.not.found"), HttpStatus.NOT_FOUND)
        }
        return deque.peekLast()
    }

    fun getTickers(markets: List<String>): List<TickerDto> =
        markets.map { getTicker(it) }

    private fun updateTickerListCache(code: String, dto: TickerDto) {
        val deque = tickerListCache.computeIfAbsent(code) { ConcurrentLinkedDeque() }

        synchronized(deque) {
            var lastOld: TickerDto? = null
            var current: TickerDto?
            val now = System.currentTimeMillis()
            val cutoff = now - SLIDING_WINDOW_MS

            while (true) {
                current = deque.pollFirst() ?: break
                if (current.timestamp < cutoff) {
                    lastOld = current
                } else {
                    deque.addFirst(current)
                    break
                }
            }
            lastOld?.let { deque.addFirst(it) }
            deque.addLast(dto)
        }
    }

    fun calculateOneMinuteRate(code: String, currentTradePrice: Double): Double {
        val deque = tickerListCache[code]
        if (deque.isNullOrEmpty()) return 0.0

        val target = System.currentTimeMillis() - SLIDING_WINDOW_MS

        return deque.filter { it.timestamp <= target }
            .maxByOrNull { it.timestamp }
            ?.let { (currentTradePrice - it.tradePrice) / it.tradePrice }
            ?: 0.0
    }

    fun pushMessage(dto: TickerDto) {
        val market = dto.code
        val now = System.currentTimeMillis()
        val mutex = pushMutexMap.computeIfAbsent(market) { Mutex() }

        coroutineScope.launch {
            mutex.withLock {
                val lastSent = lastPushSentTime[market] ?: 0L
                if (now - lastSent < PUSH_THROTTLE_INTERVAL_MS) return@withLock

                val absRate = kotlin.math.abs(dto.oneMinuteRate)
                val direction = if (dto.oneMinuteRate > 0) "HIGH" else "LOW"
                val messageKey = if (dto.oneMinuteRate > 0) "high.rate" else "low.rate"

                OneMinuteRate.entries
                    .filter { it != OneMinuteRate.NONE && absRate >= it.threshold }
                    .forEach { rate ->
                        val topic = "$market-$direction-${rate.name}"
                        val body = String.format(messageUtil.resolveMessage(messageKey), rate.threshold * 100)

                        pushService.sendAsync(
                            title = market,
                            body = body,
                            marketCode = market,
                            topic = topic
                        )
                    }

                lastPushSentTime[market] = now
            }
        }
    }

    fun publish(dto: TickerDto) {
        val market = dto.code
        val now = System.currentTimeMillis()
        val lastSent = lastSentTime[market] ?: 0L

        if (now - lastSent >= THROTTLE_INTERVAL_MS) {
            eventPublisher.publish(dto)
            lastSentTime[market] = now
        }
    }
}
